core/scripts.py

```diff
@@ -2758,6 +2766,166 @@ def suppress_work(self, library: Library, identifier: Identifier) -> None:
+        parser = argparse.ArgumentParser()
+        if _db is None:
+            raise ValueError("No database session provided.")
+
+        return parser
+
+    @classmethod
+    def parse_command_line(
+        cls, _db: Session | None = None, cmd_args: list[str] | None = None
+    ):
+        parser = cls.arg_parser(_db)
+        return parser.parse_known_args(cmd_args)[0]
+
+    def do_run(self, cmd_args: list[str] | None = None) -> None:
+        parsed = self.parse_command_line(self._db, cmd_args=cmd_args)
+
+        while True:
+            task = start_next_task(self._db, AsyncTaskType.INVENTORY_REPORT)
+            if not task:
+                return
+
+            self.process_task(task)
+
+    def process_task(self, task: AsyncTask):
+        data = InventoryReportTaskData(**task.data)
+        files = []
+        try:
+            current_time = datetime.datetime.now()
+            date_str = current_time.strftime("%Y-%m-%d_%H:%M:%s")
+            attachments = {}
+
+            for data_source_name in self.DATA_SOURCES:
+                formatted_ds_name = data_source_name.lower().replace(" ", "_")
+                file_name = f"palace-inventory-report-{formatted_ds_name}-{date_str}"
+                # generate csv file
+                file_path = self.generate_report(
+                    data_source_name=data_source_name,
+                    library_id=data.library_id,
+                )
+                # extract contents of files and prepare in a dictionary of email attachments
+                with open(file_path) as csv_file:
+                    attachments[f"{file_name}.csv"] = csv_file.read()
+                    files.append(csv_file)
+
+            self.services.email.send_email(
+                subject=f"Inventory Report {current_time}",
+                receivers=[data.admin_email],
+                text="",
+                attachments=attachments,
+            )
+            task.complete()
+        except Exception as e:
+            # log error
+            self.log.error(f"Failed to process task: {task}", e)
+            task.fail(failure_details=f"{e}")
+        finally:
+            self._db.commit()
+            for file in files:
+                os.remove(file.name)
+
+    def generate_report(self, data_source_name: str, library_id: int) -> str:
+        """Generate a csv file and return the file path"""
+        with tempfile.NamedTemporaryFile(
+            "w",
+            delete=False,
+        ) as temp:
+            writer = csv.writer(temp, delimiter=",")
+            rows = self._db.execute(
+                text(self.inventory_report_query()),
+                {"data_source_name": data_source_name, "library_id": library_id},
+            )
+            writer.writerow(rows.keys())
+            writer.writerows(rows)
+        return temp.name
+
+    def inventory_report_query(self) -> str:
+        return """
+            select
```

Comment:
This SQL statement is big.  Could it be written using the sqlalchemy api?  Surely.  Would it be more or less readable?  I would argue that it would be less readable.  Also building the query by directly talking to the database allowed me to iterate quickly in psql rather than trying to do it iteratively in python.  So IMHO making modifications to it will be relatively easy because they can be pasted into a `psql` terminal using production data.

Reply:
FYI: one of the things that I don't like about sprinkling text SQL queries into our code base, is that it makes it hard when refactoring our SQLAlchemy models because the IDE can no longer tell you where particular attributes are used. 

Your point is valid here though, and I'm willing to defer to your judgment that this is the best way to approach this query.